core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MergeProjectWithValues.java

```diff
@@ -0,0 +1,179 @@
+                    .capturedAs(VALUES)));
+
+    private final Metadata metadata;
+
+    public MergeProjectWithValues(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isMergeProjectWithValues(session);
+    }
+
+    @Override
+    public Result apply(ProjectNode node, Captures captures, Context context)
+    {
+        ValuesNode valuesNode = captures.get(VALUES);
+
+        // handle projection which prunes all symbols
+        if (node.getOutputSymbols().isEmpty()) {
+            return Result.ofPlanNode(new ValuesNode(valuesNode.getId(), valuesNode.getRowCount()));
+        }
+
+        // fix iteration order over ProjectNode's assignments
+        Map<Symbol, Expression> assignments = ImmutableMap.copyOf(node.getAssignments().getMap());
+
+        // handle values with no output symbols
+        if (valuesNode.getOutputSymbols().isEmpty()) {
+            return Result.ofPlanNode(new ValuesNode(
+                    valuesNode.getId(),
+                    node.getOutputSymbols(),
+                    nCopies(valuesNode.getRowCount(), new Row(ImmutableList.copyOf(assignments.values())))));
+        }
+
+        // do not proceed if ValuesNode contains a non-deterministic expression and it is referenced more than once by the projection
+        Set<Symbol> nonDeterministicValuesOutputs = new HashSet<>();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            Row row = (Row) rowExpression;
+            for (int i = 0; i < valuesNode.getOutputSymbols().size(); i++) {
+                if (!isDeterministic(row.getItems().get(i), metadata)) {
+                    nonDeterministicValuesOutputs.add(valuesNode.getOutputSymbols().get(i));
+                }
+            }
+        }
+        Set<Symbol> multipleReferencedSymbols = node.getAssignments().getExpressions().stream()
+                .map(SymbolsExtractor::extractAll)
+                .flatMap(Collection::stream)
+                .collect(toImmutableMultiset()).entrySet().stream()
+                .filter(entry -> entry.getCount() > 1)
+                .map(Multiset.Entry::getElement)
+                .collect(toImmutableSet());
+        if (!Sets.intersection(nonDeterministicValuesOutputs, multipleReferencedSymbols).isEmpty()) {
+            return Result.empty();
+        }
+
+        // inline values expressions into projection's assignments
+        ImmutableList.Builder<Expression> projectedRows = ImmutableList.builder();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            Map<SymbolReference, Expression> mapping = buildMappings(valuesNode.getOutputSymbols(), (Row) rowExpression);
+            Row projectedRow = new Row(assignments.values().stream()
```

Comment:
The assignments need to be iterated in the order of the projection's output symbols, since we're using that ordering below to create the ValuesNode. There's no guarantee in which order the symbols will come out of Assignments.